content/blog_posts/articles/hmr-in-adonisjs.md

```diff
@@ -0,0 +1,106 @@
+We won't go deeply in the details of the implementation, but here's a summary of how it works: 
+
+- Hot Hook is a [hook](https://nodejs.org/api/module.html#customization-hooks) for Node.js. In short: a hook is a way to execute some code when you import a module in your own codebase.
+- When you start your server with `node ace serve --hmr`, Hot Hook is loaded and will build a dependency graph of your application.
+- When you modify a file, Hot Hook will detect the change and will increment a version number for the file.
+- When the updated file is imported again ( because of a request, for example ), Hot Hook will append the version number to the `import` statement. That means we gonna `import('#controllers/users_controller.ts?version=2')`, and by doing this, we bypass the cache of the module and get the latest version of the file.
+
+This is a quick summary of how Hot Hook works, there is a bit more to it, but this is the main idea.
+
+Feel free to dive into the [Hot-Hook README](https://github.com/Julien-R44/hot-hook) for more information.
+
+## Using HMR in AdonisJS
+
+To use HMR in AdonisJS, you'll first need to install version `6.x.x` of `@adonisjs/core`.
+
+Then, also install `hot-hook`:
+
+```bash
+npm install -D hot-hook
+```
+
+Once done, you need to add the following configuration to your `package.json`
+
+```jsonc
+// title: package.json
+{
+  "name": "adonis-app",
+  "type": "module",
+  // ...
+  "hotHook": {
+    "boundaries": [
+      "./app/controllers/**/*.ts"
+    ]
+  }
+}
+```
+
+Then run `node ace serve --hmr` to start the dev server. And that's it! Try modifying a controller and you'll see that the server does not restart, and if you execute a request, the changes will be taken into account: you'll always have the latest version of your code.
+
+For more information on using it in AdonisJS, you can check out the [documentation page](https://docs.adonisjs.com/hot-module-reloading).
+
+## Why HMR?
+
+Some might ask, why add HMR to AdonisJS? Here are a few reasons why we decided to add this feature.
+
+### TSX as a template engine
+
+There's increasing talk about [using TSX as a template engine](https://adonisjs.com/blog/use-tsx-for-your-template-engine) in AdonisJS, as an alternative to Edge.
+
+With Edge, we don't need HMR, because Edge files are not JavaScript modules, they are not directly `import`ed and thus are not cached. Basically, to render an Edge view, we read the `.edge` file with `readFile` then compile it. Each time. So, we always have the latest version of the file when a view is rendered.
+
+However, with TSX: we're working with JavaScript modules. And, when a JavaScript module is imported, it gets cached. No matter how many times you modify the file, the imported module will always remain the same until the server is restarted.
+
+As a result, even a simple modification of padding/margin in our TSX view would cause a complete server restart, which can be long and tedious when developing your UI, for example.
```

Comment:
```suggestion
As a result, even a simple modification in our TSX view would cause a complete server restart, which can be long and tedious when developing your UI.
```